common/ManagementTestShared/MgmtRecordTestBase.cs

```diff
@@ -0,0 +1,224 @@
+        /// and will be compiled out of release builds to help make that easier
+        /// to spot.
+        /// </summary>
+        public bool SaveDebugRecordingsOnFailure { get; set; } = false;
+#endif
+
+        protected MgmtRecordTestBase(bool isAsync) : this(isAsync, RecordedTestUtilities.GetModeFromEnvironment())
+        {
+        }
+
+        protected MgmtRecordTestBase(bool isAsync, RecordedTestMode mode) : base(isAsync)
+        {
+            Sanitizer = new RecordedTestSanitizer();
+            Matcher = new RecordMatcher();
+            Mode = mode;
+            TestEnvironment = new TEnvironment();
+            TestEnvironment.Mode = Mode;
+            CleanupPolicy = new ResourceGroupCleanupPolicy();
+        }
+
+        private string GetSessionFilePath()
+        {
+            TestContext.TestAdapter testAdapter = TestContext.CurrentContext.Test;
+
+            string name = new string(testAdapter.Name.Select(c => s_invalidChars.Contains(c) ? '%' : c).ToArray());
+            string additionalParameterName = testAdapter.Properties.ContainsKey(ClientTestFixtureAttribute.RecordingDirectorySuffixKey) ?
+                testAdapter.Properties.Get(ClientTestFixtureAttribute.RecordingDirectorySuffixKey).ToString() :
+                null;
+
+            string className = testAdapter.ClassName.Substring(testAdapter.ClassName.LastIndexOf('.') + 1);
+
+            string fileName = name + (IsAsync ? "Async" : string.Empty) + ".json";
+            return Path.Combine(TestContext.CurrentContext.TestDirectory,
+                "SessionRecords",
+                additionalParameterName == null ? className : $"{className}({additionalParameterName})",
+                fileName);
+        }
+
+        protected abstract void InitializeClients();
+
+        protected virtual Task OnOneTimeSetupBaseAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+        protected virtual Task OnOneTimeSetupScenarioAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+        protected virtual Task OnSetupBaseAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+        protected virtual Task OnSetupScenarioAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+        protected virtual Task OnOneTimeTearDownBaseAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+        protected virtual Task OnOneTimeTearDownScenarioAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        protected virtual Task OnTearDownBaseAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        protected virtual Task OnTearDownScenarioAsync()
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        [OneTimeSetUp]
+        public void RunOneTimeSetup()
+        {
+            if (Mode == RecordedTestMode.Live || Debugger.IsAttached)
```

Comment:
I see same code on new logger() in RecordedTestBase OneTimeSetup. Since it is not virtual, we are doing it twice?